src/Search/ReverseGeocoder_P.cpp

```diff
@@ -258,66 +267,159 @@ QVector<std::shared_ptr<const OsmAnd::ReverseGeocoder::ResultEntry>> OsmAnd::Rev
+            return o1->dist < o2->dist;
+        });
+
+    if (list.size() > 0)
+    {
+        std::shared_ptr<RouteSegmentPoint> ps = nullptr;
+        //if (ctx->publicTransport)
+        //{
+        //}
+        if (ps == nullptr)
+        {
+            ps = list[0];
+        }
+        ps->others = list;
+        return ps;
+    }
+    return nullptr;
+}
+
+QVector<std::shared_ptr<const OsmAnd::ReverseGeocoder::ResultEntry>> OsmAnd::ReverseGeocoder_P::reverseGeocodeToRoads(
+        const LatLon searchPoint, const std::shared_ptr<RoutingContext>& ctx, const bool allowEmptyNames) const
+{
+    QVector<std::shared_ptr<const ResultEntry>> lst{};
+    std::vector<std::shared_ptr<RouteSegmentPoint>> listR = {};
+    this->regionsCache = {};
+    // we allow duplications to search in both files for boundary regions
+    // here we use same code as for normal routing, so we take into account current profile and sort by priority & distance
+    findRouteSegment(searchPoint, ctx, listR);
     double distSquare = 0;
-    QSet<ObfObjectId> set{};
-    QSet<QString> streetNames{};
-    for (auto p : roads)
+    QHash<QString, QList<QString>> streetNames{};
+    
+    for (std::shared_ptr<RouteSegmentPoint> p : listR)
     {
-        auto road = p.first;
-        double roadDistSquare = p.second->distSquare;
-        if (set.contains(road->id))
-            continue;
+        std::shared_ptr<RouteDataObject>& road = p->getRoad();
+        double roadDistSquare = std::pow(p->dist, 2);
+        string name;
+        if (road->getName().empty())
+        {
+            std::string langEmptyString = std::string();
+            name = road->getRef(langEmptyString, false, true);
+        }
         else
-            set.insert(road->id);
-        if (!road->captions.isEmpty())
+        {
+            name = road->getName();
+        }
+        
+        if (allowEmptyNames || !name.empty())
         {
             if (distSquare == 0 || distSquare > roadDistSquare)
+            {
                 distSquare = roadDistSquare;
-            std::shared_ptr<ResultEntry> entry = std::make_shared<ResultEntry>();
-            entry->road = road;
-            entry->streetName = road->getCaptionInNativeLanguage();
-            if (entry->streetName.isEmpty())
+            }
+            
+            std::shared_ptr<ResultEntry> sr = std::make_shared<ResultEntry>();
+            sr->searchPoint = searchPoint;
+            sr->streetName = name.empty() ? QStringLiteral() : QString::fromUtf8(name.c_str());
+            sr->point = p;
+            sr->connectionPoint = LatLon(Utilities::get31LatitudeY(p->preciseY), Utilities::get31LongitudeX(p->preciseX));
+            //sr->regionFP = road->region->filePointer;
+            //sr->regionLen = road->region->length;
+            
+            QList<QString> plst = streetNames[sr->streetName];
+            if (plst.empty())
```

Comment:
No need it here, empty key already inserted above
https://doc.qt.io/qt-5/qhash.html#operator-5b-5d